sink/datadog.go

```diff
@@ -92,29 +95,155 @@ func NewDatadog(monitorId string, opts, tags map[string]string, httpClient *http
 				return nil, err
 			}
 			d.tr = tr
+
 		case "metric-prefix":
 			if v == "" {
 				return nil, fmt.Errorf("datadog sink metric-prefix is empty string; value required when option is specified")
 			}
 			d.prefix = v
 
+		case "use-dogstatsd":
+			d.useDogStatsD = blip.Bool(v)
+
+		case "dogstatsd-host":
+			d.dogStatsDHost = v
+
+		case "dogstatsd-port":
+			d.dogStatsDPort = v
+
 		default:
 			return nil, fmt.Errorf("invalid option: %s", k)
 		}
 	}
 
-	if d.apiKeyAuth == "" {
-		return nil, fmt.Errorf("datadog sink required either api-key-auth or api-key-auth-file")
-	}
+	if d.useDogStatsD {
+		if d.dogStatsDHost == "" || d.dogStatsDPort == "" {
+			return nil, fmt.Errorf("dogStatsD host and port are required when DogStatsD is enabled")
+		}
+
+		client, err := statsd.New(fmt.Sprintf("%s:%s", d.dogStatsDHost, d.dogStatsDPort))
+		if err != nil {
+			return nil, err
+		}
+		d.dogStatsDClient = client
+	} else {
+		if d.apiKeyAuth == "" {
+			return nil, fmt.Errorf("datadog sink required either api-key-auth or api-key-auth-file")
+		}
 
-	if d.appKeyAuth == "" {
-		return nil, fmt.Errorf("datadog sink required either app-key-auth or app-key-auth-file")
+		if d.appKeyAuth == "" {
+			return nil, fmt.Errorf("datadog sink required either app-key-auth or app-key-auth-file")
+		}
+
+		c := datadog.NewConfiguration()
+		c.HTTPClient = httpClient
+		metricsApi := datadogV2.NewMetricsApi(datadog.NewAPIClient(c))
+		d.metricsApi = metricsApi
 	}
 
 	return d, nil
 }
 
 func (s *Datadog) Send(ctx context.Context, m *blip.Metrics) error {
+	if s.useDogStatsD {
+		return s.sendDogStatsD(ctx, m)
+	}
+	return s.sendApi(ctx, m)
+}
+
+func (s *Datadog) sendDogStatsD(ctx context.Context, m *blip.Metrics) error {
+	status.Monitor(s.monitorId, s.Name(), "sending metrics")
+
+	// On return, set monitor status for this sink
+	n := 0
+	defer func() {
+		status.Monitor(s.monitorId, s.Name(), "last sent %d metrics at %d", n, time.Now())
+	}()
+
+	// Pre-alloc DogStatsD data points
```

Comment:
None of this block is used. Looks like Dogstatd fires off every metric when one by one?

Reply:
Yeah you are right, I was also hoping there will be some kind of batching API but there's none. The go client does buffering by default and even does some basic client-side aggregation by default as well. I will fix it.